### PR TITLE
create chainspec struct

### PR DESCRIPTION
**Pull request by blasrodri**

### CODE REVIEW

File: src/models/chainspec.rs [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+
+#[derive(Debug, Deserialize)]
+struct Engine {
+    name: String,
+    params: EngineParams,
+}
+
+#[derive(Debug, Deserialize)]
+struct EngineParams {
+    period: f32,
```
Comment (vorot93) [inline src/models/chainspec.rs:29]: TBH not too happy about using floats. Better change format to have millisecs here. Field should be `Duration`

File: src/models/chainspec.rs [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+    epoch: f32,
```
Comment (vorot93) [inline src/models/chainspec.rs:30]: `u64`, remove decimal points from format

File: src/models/chainspec.rs [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+#[derive(Debug, Deserialize)]
+struct EngineGenesis {
+    vanity: H256,
+    signers: Vec<H160>,
+}
+
+#[derive(Debug, Deserialize)]
+struct HardForks {
+    #[serde(flatten)]
+    eips: HashMap<String, String>,
```
Comment (vorot93) [inline src/models/chainspec.rs:43]: EIPs are pretty specific, so need a struct with each eip as a separate field.

File: src/models/chainspec.rs [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+    max_code_size: String,
```
Comment (vorot93) [inline src/models/chainspec.rs:44]: Actually, this, and everywhere else, is `u64`. `primitive_types::U64` has the code to parse hex strings.

File: src/models/chainspec.rs [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+        "0x00000000000000000000000000000000000000f9" = "0x1"
+        "0x00000000000000000000000000000000000000fa" = "0x1"
+        "0x00000000000000000000000000000000000000fb" = "0x1"
+        "0x00000000000000000000000000000000000000fc" = "0x1"
+        "0x00000000000000000000000000000000000000fd" = "0x1"
+        "0x00000000000000000000000000000000000000fe" = "0x1"
+        "0x00000000000000000000000000000000000000ff" = "0x1"
+        "0x31b98d14007bdee637298086988a0bbd31184523" = "0x200000000000000000000000000000000000000000000000000000000000000"
+        "#;
+        let chain_spec = toml::from_slice::<ChainSpec>(chain_spec_str.as_ref()).unwrap();
```
Comment (vorot93) [inline src/models/chainspec.rs:459]: Must be equal with manually-filled struct

Comment (vorot93) [inline src/models/chainspec.rs:459]: Must be equal to manually-filled struct.

File: src/models/chainspec.rs [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+        max_code_size = "0x0"
+        
+        [params]
+        account_start_nonce = "0x0"
+        chain_id = "0x4"
+        gas_limit_bound_divisor = "0x400"
+        max_code_size = "0x6000"
+        maximum_extra_data_size = "0xffff"
+        min_gas_limit = "0x1388"
+        network_id = "0x4"
```
Comment (JekaMas) [inline src/models/chainspec.rs:144]: @vorot93 can we have something a bit more specific than a string for hex encoded decimals?

Comment (vorot93) [reply src/models/chainspec.rs:144]: Yeah, it should be `u64`, or `primitive_types::U64` since it has relevant decoding code.

Comment (JekaMas) [reply src/models/chainspec.rs:144]: How do you think, using u64 or primitive_types::U64 will make this ambiguous? Both u64 and string representation will be possible.

File: src/models/chainspec.rs [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+
+    fn visit_f64<E>(self, v: f64) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(v as u64)
+    }
+}
+
+fn deserialize_float_as_u64<'de, D>(deserializer: D) -> Result<u64, D::Error>
```
Comment (vorot93) [inline src/models/chainspec.rs:141]: Remove, also move to use integers in Rinkeby reference itself.

File: src/models/chainspec.rs [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+struct ChainSpec {
+    name: String,
+    data_dir: String,
+    bootnodes: Vec<NodeUrl>,
+    engine: Engine,
+    hardforks: HardForks,
+    params: Params,
+    genesis: Genesis,
+    precompiles: HashMap<String, HashMap<String, Precompiles>>,
+    balances: HashMap<H160, String>,
```
Comment (vorot93) [inline src/models/chainspec.rs:18]:
```suggestion
    balances: HashMap<Address, U256>,
```

File: src/models/chainspec.rs (hunk 7 above)
Comment (blasrodri) [reply src/models/chainspec.rs:141]: Sorry, I don't really understand what you mean here by "remove". How will this conversion be handled?

Comment (vorot93) [reply src/models/chainspec.rs:141]: There should be no conversion at all, please edit Rinkeby spec to e.g. replace `3000.0` with `3000`. I don't think we have any values with non-zero decimal places anywhere in the specs.

File: src/models/chainspec.rs [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+
+#[derive(Debug, Deserialize, PartialEq)]
+struct EngineGenesis {
+    vanity: H256,
+    signers: Vec<H160>,
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+struct HardForks {
+    eip140: U64,
```
Comment (vorot93) [inline src/models/chainspec.rs:44]: Here and below: is it possible to keep standard Rust `u64`? You could try `deserialize_with` to parse into `primitive_types::U64` and from there extract `u64`.

File: src/models/chainspec.rs [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+        name = "blake2_f"
+        pricing = { formula = "blake2_f", params = { gas_per_round = 1.0 } }
+        
+        [balances]
+        "0x0000000000000000000000000000000000000000" = "0x1"
+        "0x31b98d14007bdee637298086988a0bbd31184523" = "0x200000000000000000000000000000000000000000000000000000000000000"
+        "#;
+        let chain_spec = toml::from_slice::<ChainSpec>(chain_spec_str.as_ref()).unwrap();
+        let precompiles_0 = vec![(
+            "0x0000000000000000000000000000000000000001".into(),
```
Comment (vorot93) [inline src/models/chainspec.rs:269]:
```suggestion
            H160(hex!("0x0000000000000000000000000000000000000001")),
```

Comment (vorot93) [reply src/models/chainspec.rs:269]: Here and below

**Review by vorot93 — APPROVED** [review]
Thanks!

**Pull request merged**